### PR TITLE
Add prost, serde, tonic and crate plugins

### LINKED ISSUE #96: Plugin request for Buf Schema Registry - prost plugins
Do we want any of these?

https://buf.build/prost/plugins

- prost
- serde
- tonic
- crate

### PR DESCRIPTION
**Pull request by mfridman**
Fix #96 

- [x] License information, need to resolve this
- [x] Add serde, tonic and ~crate~[^1] plugins
- [ ] Figure out the name of the plugin

[^1]: Note, we are not going to add `prost-crate` just yet, see discussion below. TL;DR to use this plugin effectively it requires access to local fs, and although there is a flag that disables this behaviour it's unclear if it's usable in that capacity.

### CODE REVIEW

**Comment by mfridman** [comment]
From what I can see the `protoc-gen-prost-serde` and `protoc-gen-tonic` plugins rely on insertion points? Which means they cannot operate independently without the `protoc-gen-prost` plugin.

This makes it incompatible with our testing strategy.

Furthermore, the `protoc-gen-prost-crate` plugin requires a Cargo manifest template, i.e., Cargo.toml file, which opens another question on what to do with plugins that have a dependency on the filesystem. To bypass this behaviour can pass the  option:

```
opt:
      - no_features
```

This breaks another assumption that a plugin has sane defaults that can work out-of-the-box.

cc @pkwarren we should probably discuss this one.

**Comment by mfridman** [comment]
Not suggesting we do this, but after a bit of thought this _might_ be one approach:

### Handling plugins with insertion points

This applies to `protoc-gen-prost-serde` and `protoc-gen-tonic`.

a. introduce a new key in `buf.gen.yaml` that records the plugin dependency and ordering. Unlike `deps` these are plugins that must be run together and executed in the right order, maybe `insertion_points_from` key:

```
protoc-gen-prost -> protoc-gen-prost-serde
```

We can still record the `deps` and use that to resolve the DAG order. It also feels like we want to surface this somehow because an end-user looking at the plugins page alone may be confused why the plugin doesn't work. I'd go as far as validate a plugin like this before invoking generation and spitting out an error:

```
failed remote generation: when specifying protoc-gen-prost-serde plugin, you must also include protoc-gen-prost plugin and it must be generated in the correct order:

- protoc-gen-prost
- protoc-gen-prost-serde
```

b. do not modify `buf.gen.yaml`, and do not test the plugin. Not ideal, but tossing it as an option.

### Handling plugins with options

The `protoc-gen-prost-crate` plugin requires filesystem access, and so it can't function as a **remote plugin** as-is.

But, it does "work" if an option is passed such as `no_features`.

a. introduce a new `buf.gen.yaml` key to capture the "required_remote_option" option. Although technically this is not a required option in local generation.
b. ask upstream to change the behaviour, although this is very unlikely. 
c. do not support this plugin.

**Comment by pkwarren** [comment]
> ### Handling plugins with insertion points

Maybe we change `deps:` in the config to support an optional `insertion_points` key:

```yaml
deps:
  - plugin: buf.build/community/neoeinstein-protoc-gen-prost
    insertion_points: true
```

If true, when testing we always run dependencies in addition to the plugin itself.

> ### Handling plugins with options
> The `protoc-gen-prost-crate` plugin requires filesystem access, and so it can't function as a **remote plugin** as-is.

I think I'd keep this as local only for now - I don't see it being used nearly as often (it sounds like it does one time crate initialization). The buf.gen.yaml example also lists running it as a local plugin: https://github.com/neoeinstein/protoc-gen-prost#example-bufgenyaml

File: plugins/community/neoeinstein-prost/v0.2.1/buf.plugin.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+version: v1
+name: buf.build/community/neoeinstein-prost
+plugin_version: v0.2.1
+source_url: https://crates.io/crates/protoc-gen-prost
+description: Generates code using the Prost! code generation engine.
+# TODO(mf): repository has no license, but crates.io does. Is it even correct?
```
[the remark this answers is not in the document]
Comment (mfridman) [reply plugins/community/neoeinstein-prost/v0.2.1/buf.plugin.yaml:6]: https://github.com/bufbuild/plugins/issues/96#issuecomment-1343808232

**Comment by mfridman** [comment]
Okay, I cleaned up most of this PR. The outstanding issue is the testing strategy we employed.

There was an assumption we could package / build / test a single plugin in isolation.

But for tonic and serde, that is not the case because those plugins don't produce output _on their own_, so we have nothing to checksum.

**Comment by mfridman** [comment]
Added `protoc-gen-prost` in https://github.com/bufbuild/plugins/pull/288. Going to close this PR and open a separate one for serde and tonic specifically.

**Pull request closed without merging**